home/api_v3.py

```diff
@@ -103,74 +104,58 @@ class ContentPagesV3APIViewset(PagesAPIViewSet):
+            if (
+                "return_drafts" in request.GET
+                and self.request.query_params.get("return_drafts", "").lower() == "true"
+            ):
                 instance = self.get_object().get_latest_revision_as_object()
+
             else:
                 instance = self.get_object()
 
-            instance.save_page_view(request.query_params)
-            serializer = ContentPageSerializerV3(instance, context={"request": request})
-            return Response(serializer.data)
-
-        except Exception:
+        except Http404:
             raise NotFound({"page": ["Page matching query does not exist."]})
 
-        return super().detail_view(request, pk)
+        instance.save_page_view(request.query_params)
+        serializer = ContentPageSerializerV3(instance, context={"request": request})
+        return Response(serializer.data)
 
     def listing_view(self, request, *args, **kwargs):
-        # If this request is flagged as QA then we should display the pages that have the filtering tags
-        # or triggers in their draft versions
-        if "qa" in request.GET and request.GET["qa"] == "True":
-            tag = self.request.query_params.get("tag")
-            trigger = self.request.query_params.get("trigger")
-            have_new_triggers = []
-            have_new_tags = []
-            # have_new_buttons = []
-            unpublished = ContentPage.objects.filter(has_unpublished_changes="True")
-
-            for page in unpublished:
-                latest_rev = page.get_latest_revision_as_object()
-
-                if trigger and latest_rev.triggers.filter(name=trigger).exists():
-                    have_new_triggers.append(page.id)
-                if tag and latest_rev.tags.filter(name=tag).exists():
-                    have_new_tags.append(page.id)
-
-            queryset = self.get_queryset()
-
-            self.check_query_parameters(queryset)
-            queryset = self.filter_queryset(queryset)
-            queryset = queryset | ContentPage.objects.filter(id__in=have_new_triggers)
-            queryset = queryset | ContentPage.objects.filter(id__in=have_new_tags)
-
-            queryset_list = self.paginate_queryset(queryset)
-
-            serializer = ContentPageSerializerV3(
-                queryset_list, context={"request": request}, many=True
-            )
-            return self.get_paginated_response(serializer.data)
-
         queryset = self.get_queryset()
+
+        channel = ""
+        if "channel" in self.request.query_params:
+            channel = self.request.query_params.get("channel", "").lower()
+
+            if channel in {"web", "whatsapp", "sms", "ussd", "messenger", "viber"}:
+                queryset = queryset.filter(**{f"enable_{channel}": True})
+            else:
+                raise ValidationError(
+                    {"channel": [f"Channel matching query '{channel}' does not exist."]}
+                )
```

Comment:
I think we can factor out the channel lookup and validation check into a function that we call anywhere we need to get the channel:

```python
channel = get_channel(self.request.query_params)
if channel:
    queryset = queryset.filter(**{f"enable_{channel}": True})
```